lib/dynamic_type/src/dynamic_type/dynamic_type.h

```diff
@@ -588,151 +610,153 @@ DEFINE_BINARY_OP(rshift, >>);
-    });                                                                        \
-    if (ret.has_value()) {                                                     \
-      return ret.value();                                                      \
-    }                                                                          \
-    if constexpr (opcheck<LHS>.hasExplicitCastTo(opcheck<DT>)) {               \
-      return (DT)x op y;                                                       \
-    }                                                                          \
-    DYNAMIC_TYPE_CHECK(                                                        \
-        false,                                                                 \
-        "Cannot compute ",                                                     \
-        typeid(LHS).name(),                                                    \
-        " ",                                                                   \
-        #op,                                                                   \
-        " ",                                                                   \
-        y.type().name(),                                                       \
-        " : incompatible type");                                               \
+#define DEFINE_COMPARE_OP(opname, op)                                      \
+  /*TODO: we should inline the definition of lambdas into enable_if,*/     \
+  /*but I can only do this in C++20 */                                     \
+  constexpr auto opname##_defined_checker = [](auto x, auto y) constexpr { \
+    using X = typename decltype(x)::type;                                  \
+    using Y = typename decltype(y)::type;                                  \
+    if constexpr (opcheck<X> op opcheck<Y>) {                              \
+      return std::is_convertible_v<                                        \
+          decltype(std::declval<X>() op std::declval<Y>()),                \
+          bool>;                                                           \
+    }                                                                      \
+    return false;                                                          \
+  };                                                                       \
+  template <                                                               \
+      typename DT,                                                         \
+      typename = std::enable_if_t<                                         \
+          is_dynamic_type_v<DT> &&                                         \
+          any_check(                                                       \
+              opname##_defined_checker,                                    \
+              DT::type_identities_as_tuple_maybe_no_container,             \
+              DT::type_identities_as_tuple_maybe_no_container)>>           \
```

Comment:
The only change is replacing `type_identities_as_tuple` with `type_identities_as_tuple_maybe_no_container`, no other changes.